Handpose interface

ziyuan-linn opened the pull request:
Hi everyone, I implemented a new handpose API according to this discussion #22. 
- add `detect()`, `detectStart()`, and `detectStop()` function
- remove EventEmitter interface
- add support for p5 preload to `ml5.handpose()` function
- add examples for single image detection and start/stop functionality

Some issues that came up during implementation: 

The current `ml5.handpose` function returns an `Handpose` object in order to support p5 preload. Thus it currently cannot support the promise interface. There might be a workaround to support both interfaces, but I will reach out to the p5 team and get a better grasp of `preload()` first.

If `detectStart()` is called twice, for instance:
```javascript
detectStart(video, gotHands);
detectStart(video2, gotHands);
```
The model would not output any warning and output detection for video2 only. Please let me know if this behavior makes sense.

To-Dos:
- figure out if both p5 preload and promise interface can be supported
- implement named landmark output

shiffman commented on examples/Handpose-p5-image/sketch.js:
```diff
@@ -0,0 +1,39 @@
+  noStroke();
+  // Draw the image
+  image(img, 0, 0);
+  // start detecting hands from the webcam video
```
```suggestion
  // Detect hands in an image
```

ziyuan-linn replied:
Committing this suggestion!

shiffman commented on examples/Handpose-p5-image/sketch.js:
```diff
@@ -0,0 +1,39 @@
+// Callback function for when handpose outputs data
+function gotHands(results) {
+  // save the output to the hands variable
+  hands = results;
```
For this example, if the drawing is happening in `gotHands()` I don't think we need a global variable. I would either just draw from `results` only or move drawing to the `draw()` loop? Not sure which is a better foundational example? You could also name the argument to `gotHands()` as `hands` if you like?

ziyuan-linn replied:
Thank you for this suggestion! I would lean toward moving the drawing to the `draw()` loop since the students might want to modify the example into something more interesting and it would be easier if the drawing is already in `draw()`.

shiffman commented on examples/Handpose-p5-start-stop/sketch.js:
```diff
@@ -0,0 +1,69 @@
+function preload() {
+  // Load the handpose model.
+  handpose = ml5.handpose();
+  console.log("preload");
```
I would remove this or change it to "Hand Pose model is loaded." or something like that?

ziyuan-linn replied:
Ah, this was for testing and I totally forgot to remove it. Thank you for pointing it out, removing now!

shiffman commented on examples/Handpose-p5-start-stop/sketch.js:
```diff
@@ -0,0 +1,69 @@
+
+function setup() {
+  createCanvas(640, 480);
+  fill(0, 255, 0);
+  textSize(25);
+  noStroke();
```
I think it's generally more helpful and clear to place these where the drawing is happening (even though it's redundant). With these in `setup()`  I find it can cause confusion for students adapting the examples to include more elaborate drawing.

ziyuan-linn replied:
I have moved these to their corresponding locations!

shiffman reviewed (commented):
This is wonderful @ziyuan-linn! A few thoughts below:

1. If `startDetect()` does not support more than one video source (which is fine), we should make this clear in the documentation (cc @MOQN and @sproutleaf) and maybe output a warning if it's called twice.

2. It would be nice to include an example that shows how to to track maybe 2-3 points from the hand by name. This is often what I start people with. One idea would be to check the distance between the top point of the index (pointer?) finger and top point of the thumb. That could be mapped to the size of a circle for example? I'm happy to work on this if you like!

Other comments include inline with the code!

ziyuan-linn commented:
@shiffman thank you for your review! I made some changes according to your inline suggestions. I added a warning when `detectStart()` is called twice without a call to `detectStop()` in between. I also added named keypoints (named after the tf.js' `name` field for each landmark point). 

Please let me know if there are any other issues that need to be addressed. Otherwise, I think we can merge this so you can add the keypoint-by-name examples?

shiffman commented:
Looks great @ziyuan-linn! I'm going to go ahead and merge. If others have comments or questions, we can address them in separate issues / branches!
